Rubberduck.Core/UI/Refactorings/AnnotateDeclaration/AnnotationArgumentViewModel.cs

```diff
@@ -0,0 +1,190 @@
+                ValidateArgument();
+                OnPropertyChanged();
+            }
+        }
+
+        private void ValidateArgument()
+        {
+            var errors = ArgumentValidationErrors();
+
+            if (errors.Any())
+            {
+                SetErrors("ArgumentValue", errors);
+            }
+            else
+            {
+                ClearErrors();
+            }
+        }
+
+        private List<string> ArgumentValidationErrors()
+        {
+            var errors = new List<string>();
+
+            if (string.IsNullOrEmpty(ArgumentValue))
+            {
+                errors.Add(RubberduckUI.AnnotationArgument_ValidationError_EmptyArgument);
+            }
+            if (ArgumentValue.Length > MaxAllowedCharacters)
+            {
+                errors.Add(string.Format(RubberduckUI.AnnotationArgument_ValidationError_TooLong, MaxAllowedCharacters));
+            }
+            if (ContainsNewline(ArgumentValue))
+            {
+                errors.Add(RubberduckUI.AnnotationArgument_ValidationError_Newline);
+            }
+            else if (ContainsControlCharacter(ArgumentValue))
+            {
+                errors.Add(RubberduckUI.AnnotationArgument_ValidationError_SpecialCharacters);
+            }
+
+            switch (ArgumentType)
+            {
+                case AnnotationArgumentType.Attribute:
+                    if (!ArgumentValue.StartsWith("VB_") || ContainsWhitespace(ArgumentValue))
+                    {
+                        errors.Add(RubberduckUI.AnnotationArgument_ValidationError_AttributeNameStart);
+                    }
+                    if (ContainsWhitespace(ArgumentValue))
+                    {
+                        errors.Add(RubberduckUI.AnnotationArgument_ValidationError_WhitespaceInAttribute);
+                    }
+                    break;
+                case AnnotationArgumentType.Inspection:
+                    if (!InspectionNames.Contains(ArgumentValue))
+                    {
+                        errors.Add(RubberduckUI.AnnotationArgument_ValidationError_InspectionName);
+                    }
+                    break;
+                case AnnotationArgumentType.Boolean:
+                    if (!bool.TryParse(ArgumentValue, out _))
+                    {
+                        errors.Add(RubberduckUI.AnnotationArgument_ValidationError_NotABoolean);
+                    }
+                    break;
+                case AnnotationArgumentType.Number:
+                    if (!decimal.TryParse(ArgumentValue, out _))
+                    {
+                        errors.Add(RubberduckUI.AnnotationArgument_ValidationError_NotANumber);
+                    }
+                    break;
+            }
+
+            return errors;
+        }
+
+        private bool ContainsNewline(string argumentText)
+        {
+            return argumentText.Contains('\n') || argumentText.Contains('\r');
+        }
```

Comment:
Question/Curious: Is this equivalent to `argumentText.Contains(Environment.Newline);` or are there other considerations?

Reply:
It's not equivalent: matching `Environment.NewLine` would match `\r\n` but not `\r` and not `\n`; this matches `\r`, `\n`, and `\r\n`.

Reply:
On windows, they should not be equivalent since `Environment.Newline == "\r\n"`. That would not detect a newline copied in from a Unix style formatted text file. However, the next validation would complain about the control character.